File: src/main/java/org/distril/beengine/player/handler/InventoryPacketHandler.java
package org.distril.beengine.player.handler;

import com.nukkitx.protocol.bedrock.data.inventory.ItemStackRequest;
import com.nukkitx.protocol.bedrock.data.inventory.stackrequestactions.*;
import com.nukkitx.protocol.bedrock.handler.BedrockPacketHandler;
import com.nukkitx.protocol.bedrock.packet.*;
import lombok.extern.log4j.Log4j2;
import org.distril.beengine.inventory.transaction.ItemStackTransaction;
import org.distril.beengine.inventory.transaction.action.*;
import org.distril.beengine.network.data.transaction.ItemUseTransaction;
import org.distril.beengine.player.Player;

import java.util.Arrays;
import java.util.Collections;

@Log4j2
public class InventoryPacketHandler implements BedrockPacketHandler {

	private final Player player;

	private final ItemStackTransaction transaction;

	public InventoryPacketHandler(Player player) {
		this.player = player;
		this.transaction = new ItemStackTransaction(player);
	}

	@Override
	public boolean handle(ItemStackRequestPacket packet) {
		var responsePacket = new ItemStackResponsePacket();

		boolean continueActions = true;
		for (ItemStackRequest request : packet.getRequests()) {
			for (StackRequestActionData action : request.getActions()) {
				if (!continueActions) {
					break;
				}

				switch (action.getType()) {
					case TAKE -> {
						var takeAction = (TakeStackRequestActionData) action;

						continueActions = this.transaction.handle(new TakeItemStackAction(
								takeAction.getSource(), takeAction.getDestination(),
								this.transaction, takeAction.getCount(), request.getRequestId()));
					}

					case PLACE -> {
						var placeAction = (PlaceStackRequestActionData) action;

						continueActions = this.transaction.handle(new PlaceItemStackAction(
								placeAction.getSource(), placeAction.getDestination(),
								this.transaction, placeAction.getCount(), request.getRequestId()));
					}

					case SWAP -> {
						var swapAction = (SwapStackRequestActionData) action;

						continueActions = this.transaction.handle(new SwapItemStackAction(
								swapAction.getSource(), swapAction.getDestination(), this.transaction));
					}

					case DROP -> {
						// todo when i added player and world method for drop items
					}

					case DESTROY -> {
						var destroyAction = (DestroyStackRequestActionData) action;

						continueActions = this.transaction.handle(new DestroyItemStackAction(
								destroyAction.getSource(), destroyAction.getCount(), this.transaction));
					}

					case CRAFT_CREATIVE -> {
						var craftCreativeAction = (CraftCreativeStackRequestActionData) action;

						continueActions = this.transaction.handle(new CraftCreativeItemStackAction(craftCreativeAction.getCreativeItemNetworkId(), this.transaction));
					}

					case CRAFT_RESULTS_DEPRECATED -> {
						// skip it because deprecated
					}


					default -> log.warn("Missing inventory action handler: " + action.getType());
				}
			}

			responsePacket.getEntries().add(new ItemStackResponsePacket.Response(
					this.transaction.getStatus(),
					request.getRequestId(),
					this.transaction.getContainerEntries()
			));

			this.transaction.clear();
		}

		this.player.sendPacket(responsePacket);
		return true;
	}

	@Override
	public boolean handle(InventoryTransactionPacket packet) {
		if (this.player.isSpectator()) {
			return true;
		}

		switch (packet.getTransactionType()) {
			case ITEM_USE -> {
				var transaction = ItemUseTransaction.read(packet);

				switch (transaction.getType()) {
					case CLICK_BLOCK -> {
						player.setUsingItem(false);

						var blockPosition = transaction.getBlockPosition();
						var world = player.getWorld();
						var blockFace = transaction.getBlockFace();
						if (player.canInteract(blockPosition.toFloat().add(0.5, 0.5, 0.5))) {
							var clientItem = transaction.getItemInHand();
							var serverItem = player.getInventory().getItemInHand();
							if (player.isCreative()) {
								if (world.useItemOn(blockPosition, serverItem, blockFace, packet.getClickPosition(), player) != null) {
									return true;
								}
							} else if (serverItem.equals(clientItem)) {
								var oldServerItem = serverItem;
								if ((serverItem = world.useItemOn(blockPosition, serverItem, blockFace, packet.getClickPosition(), player)) != null) {
									if (!serverItem.equals(oldServerItem) ||
											serverItem.getCount() != oldServerItem.getCount()) {
										player.getInventory().setItemInHand(serverItem);
										player.getInventory().sendHeldItem(player.getViewers());
									}

									return true;
								}
							}
						}

						player.getInventory().sendHeldItem(Collections.singleton(player));

						if (blockPosition.distanceSquared(player.getPosition().toInt()) > 10000) {
							return true;
						}

						var target = world.getBlock(blockPosition);
						var block = target.getSide(blockFace);

						world.sendBlocks(Collections.singleton(player), Arrays.asList(target, block), UpdateBlockPacket.FLAG_ALL_PRIORITY);
						player.setUsingItem(false);
						return true;
					}

					case CLICK_AIR -> {

					}

					case BREAK_BLOCK -> {

					}
				}
			}

			case ITEM_USE_ON_ENTITY -> {

			}

			case ITEM_RELEASE -> {

			}

			default -> log.info("Unhandled transaction type: {}", packet.getTransactionType());
		}

		log.info(packet.toString());
		return true;
	}

	@Override
	public boolean handle(MobEquipmentPacket packet) {
		this.player.getInventory().setHeldItemIndex(packet.getHotbarSlot());
		return true;
	}

	@Override
	public boolean handle(ContainerClosePacket packet) {
		this.player.closeOpenedInventory();
		return true;
	}
}

File: src/main/java/org/distril/beengine/player/handler/InventoryPacketHandler.kt
package org.distril.beengine.player.handler

import com.nukkitx.protocol.bedrock.data.inventory.TransactionType
import com.nukkitx.protocol.bedrock.data.inventory.stackrequestactions.*
import com.nukkitx.protocol.bedrock.handler.BedrockPacketHandler
import com.nukkitx.protocol.bedrock.packet.*
import org.apache.logging.log4j.LogManager
import org.distril.beengine.inventory.transaction.ItemStackTransaction
import org.distril.beengine.inventory.transaction.action.*
import org.distril.beengine.material.item.ItemPalette
import org.distril.beengine.network.data.transaction.ItemUseTransaction
import org.distril.beengine.player.Player

class InventoryPacketHandler(private val player: Player) : BedrockPacketHandler {

	private val transaction = ItemStackTransaction(this.player)

	override fun handle(packet: ItemStackRequestPacket): Boolean {
		val responsePacket = ItemStackResponsePacket()
		var continueActions = true
		for (request in packet.requests) {
			for (action in request.actions) {
				if (!continueActions) break

				when (action.type) {
					StackRequestActionType.TAKE -> {
						val takeAction = action as TakeStackRequestActionData
						continueActions = this.transaction.handle(
							TakeItemStackAction(
								takeAction.source, takeAction.destination,
								this.transaction, takeAction.count.toInt(), request.requestId
							)
						)
					}

					StackRequestActionType.PLACE -> {
						val placeAction = action as PlaceStackRequestActionData
						continueActions = this.transaction.handle(
							PlaceItemStackAction(
								placeAction.source, placeAction.destination,
								this.transaction, placeAction.count.toInt(), request.requestId
							)
						)
					}

					StackRequestActionType.SWAP -> {
						val swapAction = action as SwapStackRequestActionData
						continueActions = this.transaction.handle(
							SwapItemStackAction(
								swapAction.source, swapAction.destination, this.transaction
							)
						)
					}

					StackRequestActionType.DROP -> {
						// todo when i added player and world method for drop items
					}

					StackRequestActionType.DESTROY -> {
						val destroyAction = action as DestroyStackRequestActionData
						continueActions = this.transaction.handle(
							DestroyItemStackAction(
								destroyAction.source, destroyAction.count.toInt(), this.transaction
							)
						)
					}

					StackRequestActionType.CRAFT_CREATIVE -> {
						val craftCreativeAction = action as CraftCreativeStackRequestActionData
						this.transaction.creativeOutput =
							ItemPalette.getCreativeItem(craftCreativeAction.creativeItemNetworkId)
						continueActions = this.transaction.handle(CraftCreativeItemStackAction(this.transaction))
					}

					StackRequestActionType.CRAFT_RESULTS_DEPRECATED -> {
						// skip it because deprecated
					}

					else -> log.warn("Missing inventory action handler: ${action.type}")
				}
			}

			responsePacket.entries.add(
				ItemStackResponsePacket.Response(
					this.transaction.status,
					request.requestId,
					this.transaction.containersEntries
				)
			)

			this.transaction.clear()
		}

		this.player.sendPacket(responsePacket)
		return true
	}

	override fun handle(packet: InventoryTransactionPacket): Boolean {
		if (this.player.isSpectator) return true

		when (packet.transactionType) {
			TransactionType.ITEM_USE -> {
				val transaction = ItemUseTransaction.read(packet)
				when (transaction.type) {
					ItemUseTransaction.Type.CLICK_BLOCK -> {
						this.player.setUsingItem(false)

						val blockPosition = transaction.blockPosition
						val world = player.world
						val blockFace = transaction.blockFace
						if (this.player.canInteract(blockPosition.toFloat().add(0.5, 0.5, 0.5))) {
							val clientItem = transaction.itemInHand
							var serverItem = this.player.inventory.getItemInHand()
							if (this.player.isCreative) {
								if (world.useItemOn(blockPosition, serverItem, blockFace, this.player) != null) {
									return true
								}
							} else if (serverItem == clientItem) {
								val oldServerItem = serverItem
								if (world.useItemOn(blockPosition, serverItem, blockFace, this.player).run {
										if (this != null) {
											serverItem = this
											return@run true
										}

										return@run false
									}
								) {
									if (serverItem != oldServerItem || serverItem.count != oldServerItem.count) {
										this.player.inventory.setItemInHand(serverItem)
										this.player.inventory.sendHeldItem(this.player.viewers)
									}

									return true
								}
							}
						}

						this.player.inventory.sendHeldItem(listOf(this.player))
						if (blockPosition.distanceSquared(player.position.toInt()) > 10000) return true

						val target = world.getBlock(blockPosition)
						val block = target.getSide(blockFace)
						world.sendBlocks(setOf(this.player), listOf(target, block))

						this.player.setUsingItem(false)
						return true
					}

					ItemUseTransaction.Type.CLICK_AIR -> TODO()

					ItemUseTransaction.Type.BREAK_BLOCK -> TODO()
				}
			}

			TransactionType.ITEM_USE_ON_ENTITY -> TODO()

			TransactionType.ITEM_RELEASE -> TODO()

			else -> log.info("Unhandled transaction type: ${packet.transactionType}")
		}

		log.info(packet.toString())
		return true
	}

	override fun handle(packet: MobEquipmentPacket): Boolean {
		this.player.inventory.heldItemIndex = packet.hotbarSlot
		return true
	}

	override fun handle(packet: ContainerClosePacket): Boolean {
		this.player.closeOpenedInventory()
		return true
	}

	override fun handle(packet: PacketViolationWarningPacket): Boolean {
		log.warn("Packet violation $packet")
		return true
	}

	companion object {

		private val log = LogManager.getLogger(InventoryPacketHandler::class.java)
	}
}
